src/steamship/agents/functional/functions_based.py

```diff
@@ -81,4 +75,58 @@ def next_action(self, context: AgentContext) -> Action:
         # Run the default LLM on those messages
         output_blocks = self.llm.chat(messages=messages, tools=self.tools)
 
-        return self.output_parser.parse(output_blocks[0].text, context)
+        future_action = self.output_parser.parse(output_blocks[0].text, context)
+        if not isinstance(future_action, FinishAction):
+            # record the LLM's function response in history
+            self._record_action_selection(future_action, context)
+        return future_action
+
+    def _function_calls_since_last_user_message(self, context: AgentContext) -> List[Block]:
+        function_calls = []
+        for block in context.chat_history.messages[::-1]:  # is this too inefficient at scale?
```

Comment:
This is exactly the line of code I've been wondering about too.

You think if we put some kind of optimized query into the engine/database that returns either the right subset of history or on-the-fly tags demarcate certain things like this, it could be a way to do it that was both faster and also let the python half be simpler?

Reply:
(fwiw I think it's fine to pursue the slower-but-simple solution now and then we'll optimize it after proving to ourselves it works)